R/asis-nodes.R

```diff
@@ -287,36 +287,104 @@ protect_tickbox <- function(body, ns) {
 }
 
 # FOOTNOTES --------------------------------------------------------------------
-#nocov start
-footnote_check <- function(body, ns = md_ns()) {
-  predicate <- "contains(text(), '[^') and contains(text(), ']')"
-  cascade <- glue::glue(".//md:paragraph[*[{predicate}]]")
-  xml2::xml_find_all(body, cascade, ns = ns)
-}
 
-fix_footnotes <- function(feet) {
-  char <- as.character(feet)
-  char <- gsub(
-    pattern = "([\\[][\\^]|\\])",
-    replacement = "</text><text asis='true'>\\1</text><text>",
-    x = char,
-    perl = TRUE
-  )
-  make_text_nodes(char)
+#' Protect unescaped square brackets from being escaped
+#'
+#' Commonmark allows both `[unescaped]` and `\[escaped\]` square brackets, but
+#' in the XML representation, it makes no note of which square brackets were 
+#' originally escaped and thus will escape both in the output.
+#'
+#' To fix this, we dive back into the source document and find the brackets that
+#' were originally escaped and wrap those characters in protective tags so that
+#' they do not become escaped when the XML is written. This has the benefit of 
+#' protecting footnotes and reference style links for child documents.
+#'
+#' @inheritParams resolve_anchor_links
+#' @param offset the offset in lines of the yaml header defaults to 0,
+#' indicating that the text has the yaml header stripped off. 
+#' @keywords internal
+#' @note This requires documents to be read in with `sourcepos = TRUE`
+#'
+#' @examples
+#' f <- system.file("extdata", "link-test.md", package = "tinkr")
+#' md <- yarn$new(f, sourcepos = TRUE, unescaped = FALSE)
+#' md$show()
+#' if (requireNamespace("withr")) {
+#' lnks <- withr::with_namespace("tinkr", 
+#'   protect_unescaped(md$body, readLines(md$path), offset = length(md$yaml)))
+#' md$body <- lnks
+#' md$show()
+#' }
+protect_unescaped <- function(body, txt, offset = 0L, ns = md_ns()) {
+  has_sourcepos <- xml2::xml_find_lgl(body, "boolean(.//@sourcepos)")
+  if (!has_sourcepos) {
+    warning("`protect_unescaped()` needs to know the source position. No modification taking place.")
+    return(body)
+  }
+  body <- copy_xml(body)
+  squares <- find_escaped_squares(txt)
+  escapes <- which(vapply(squares, sum, integer(1)) > 0L)
+  XPATH <- ".//md:text[not(@asis)][contains(text(), '[') or contains(text(), ']')]"
+  snodes <- xml2::xml_find_all(body, XPATH, ns = ns)
+  fix_unescaped_squares(snodes, txt, squares, escapes, offset)
+  copy_xml(body)
 }
 
-protect_footnotes <- function(body, ns = md_ns()) {
-  body <- copy_xml(body)
 
-  feet <- footnote_check(body, ns)
-  if (length(feet) == 0) {
-    return(body)
+find_escaped_squares <- function(txt) {
+  gregexpr("(\\\\\\])|(\\\\\\[)", txt)
```

Comment:
:sweat_smile:

Reply:
This took me a good five minutes to figure out how to get working 😅